Mathlib/Tactic/Linter/DeprecatedModule.lean

```diff
@@ -0,0 +1,146 @@
+
+The `deprecated.module` linter emits a warning when a file that has been renamed or split
+is imported.
+
+The usage is as follows. Write
+```lean
+import B
+...
+import Z
+
+deprecated_module since yyyy-mm-dd
+```
+in module `A` with the expectation that `A` contains nothing else.
+This triggers the `deprecated.module` linter to notify every file with `import A`
+to instead import the *direct imports* of `A`, that is `B, ..., Z`.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+The `deprecated.module` linter emits a warning when a file that has been renamed or split
+is imported.
+The default value is `true`, since this linter is designed to warn projects downstream of `Mathlib`
+of refactors and deprecations in `Mathlib` itself.
+-/
+register_option linter.deprecated.module : Bool := {
+  defValue := true
+  descr := "enable the `deprecated.module` linter"
+}
+
+/--
+Defines the `deprecatedModuleExt` extension for adding a `HashSet` of pairs of
+* a module `Name` that has been deprecated and
+* an array of `Name`s of modules that should be imported instead
+
+to the environment.
+-/
+initialize deprecatedModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := (·.foldl Std.HashSet.insertMany {})
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecatedModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores the `Init` import, since this is a special module that is expected to be imported
+by all files.
+
+It also ignores the `Mathlib.Tactic.Linter.DeprecatedModule` import (namely, the current file),
+since there is no need to import this module.
+-/
+def addModuleDeprecation {m : Type → Type} [Monad m] [MonadEnv m] [MonadQuotation m] : m Unit := do
+  let modName ← getMainModule
+  modifyEnv (deprecatedModuleExt.addEntry ·
+    (modName, (← getEnv).imports.filterMap fun i ↦
+      if i.module == `Init ||
+         i.module == `Mathlib.Tactic.Linter.DeprecatedModule then none else i.module))
+
+/--
+`deprecated_module since yyyy-mm-dd` deprecates the current module `A` in favour of
+its direct imports.
+This means that any file that directly imports `A` will get a notification on the `import A` line
+suggesting to instead import the *direct imports* of `A`.
+-/
+elab (name := deprecated_modules)
+    "deprecated_module " &"since " yyyy:num "-" mm:num "-" dd:num : command => do
+  let year := 2025
+  if yyyy.getNat < year then
+    throwErrorAt yyyy "The year should be at least {year}!"
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02, ..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02, ..., 31!"
```

Comment:
Please use `Std.Time.PlainDate.fromLeanDateString (input : String) : Except String PlainDate` (or just drop the checks?)

Reply:
Thanks: I used `Std.Time`.